pyuvdata/tests/__init__.py

```diff
@@ -5,19 +5,245 @@
+        else:
+            raise TypeError(msg % type(match))
+
+        self.expected_warning = expected_warning_list
+        self.match = match_list
+
+    @property
+    def warnlist(self):
+        """The list of recorded warnings."""
+        return self._list
+
+    def __getitem__(self, i: int):
+        """Get a recorded warning by index."""
+        return self._list[i]
+
+    def __iter__(self):
+        """Iterate through the recorded warnings."""
+        return iter(self._list)
+
+    def __len__(self):
+        """The number of recorded warnings."""
+        return len(self._list)
+
+    def pop(self, cls):
+        """Pop the first recorded warning, raise exception if not exists."""
+        for i, w in enumerate(self._list):
+            if issubclass(w.category, cls):
+                return self._list.pop(i)
+        __tracebackhide__ = True
+        raise AssertionError("%r not found in warning list" % cls)
+
+    def clear(self) -> None:
+        """Clear the list of recorded warnings."""
+        self._list[:] = []
+
+    def __enter__(self):
+        if self._entered:
+            __tracebackhide__ = True
+            raise RuntimeError("Cannot enter %r twice" % self)
+        _list = super().__enter__()
+        # record=True means it's None.
+        assert _list is not None
+        self._list = _list
+        warnings.simplefilter("always")
+        warnings.filterwarnings("ignore", message="numpy.dtype size changed")
+        warnings.filterwarnings("ignore", message="numpy.ufunc size changed")
+
+        # Filter iers warnings if iers.conf.auto_max_age is set to None, as we
+        # do in testing if the iers url is down. See conftest.py for more info.
+        if iers.conf.auto_max_age is None:
+            warnings.filterwarnings("ignore", message="failed to download")
+            warnings.filterwarnings("ignore", message="time is out of IERS range")
+
+            test_message = []
+            for message in self.match:
+                if message is None:
+                    test_message.append(False)
+                else:
+                    test_message.append(message.startswith("LST values stored in "))
+            if not any(test_message):
+                warnings.filterwarnings("ignore", message="LST values stored in ")
+
+        return self
+
+    def __exit__(
+        self, exc_type=None, exc_val=None, exc_tb=None,
+    ):
+        if not self._entered:
+            __tracebackhide__ = True
+            raise RuntimeError("Cannot exit %r without entering first" % self)
+
+        super().__exit__(exc_type, exc_val, exc_tb)
+
+        # Built-in catch_warnings does not reset entered state so we do it
+        # manually here for this context manager to become reusable.
+        self._entered = False
+
+        # only check if we're not currently handling an exception
+        if exc_type is None and exc_val is None and exc_tb is None:
+            if self.expected_warning is not None:
```

Comment:
I'm a little confused as to the use of the object in the case where `self.expected_warning is None` though. Just feels like it does nothing. Is that the behavior we want?

Reply:
It can be used to check that the right kind of warning is emitted without checking the warning message. I included it because `pytest.warns` supports that and I wanted to be able to make this be a drop in replacement. But I'm willing to take it out and error if the warning message isn't provided.

Reply:
Right, that loop happens down in line 145. But even the length checking happens inside of this if statement and since there is no `else` to this if I am just confused why we have it.

Reply:
I said the wrong thing before. This approach is used in `pytest.warns` if you don't want to assert anything about the warning but want to get back the list of warning objects so that you can do your own tests on them. I think this is a pretty infrequent use case, but I suppose it could be used to check that there are no warnings -- I need to test that usage though and make sure it actually works for that.

Reply:
Oh okay I see. thanks for the clarification.

Reply:
So it turns out I was wrong, but I think that functionality is useful, so I made some modifications so that if you pass `None` as the first argument it will assert that no warnings are raised by the code inside the context manager.